include/seqan3/io/detail/safe_filesystem_entry.hpp

```diff
@@ -0,0 +1,97 @@
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides seqan3::detail::safe_filesystem_entry.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <cassert>
+#include <system_error>
+
+#include <seqan3/core/platform.hpp>
+#include <seqan3/std/filesystem>
+
+namespace seqan3::detail
+{
+
+/*!\brief A safe guard to manage a filesystem entry, e.g. a file or a directory.
+ * \ingroup io
+ *
+ * \details
+ *
+ * This raii-wrapper class allows for a safe removal of a created filesystem entry such as a directory or file.
+ * This wrapper class assumes owning semantics. It is not copy-constructible or copy-assignable. In order to
+ * prevent misuse also the default constructor is deleted.
+ *
+ * The following example demonstrates the use case.
+ *
+ * \include test/snippet/io/detail/safe_filesystem_entry_snippet.cpp
+ */
+class safe_filesystem_entry
+{
+public:
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    safe_filesystem_entry() = delete;                                             //!< Deleted.
+    safe_filesystem_entry(safe_filesystem_entry const &) = delete;                //!< Deleted.
+    safe_filesystem_entry(safe_filesystem_entry &&) = default;                    //!< Defaulted.
+    safe_filesystem_entry & operator=(safe_filesystem_entry const &) = delete;    //!< Deleted.
+    safe_filesystem_entry & operator=(safe_filesystem_entry &&) = default;        //!< Defaulted.
+
+    /*!\brief Constructs the safe guard from a std::filesystem::path.
+     * \param p The path pointing to a filesystem entry.
+     */
+    safe_filesystem_entry(std::filesystem::path p) : entry(std::move(p))
+    {}
+
+    //!\brief Calls std::filesystem::remove_all on the wrapped entry.
+    ~safe_filesystem_entry()
+    {
+        std::error_code ec;
+        std::filesystem::remove_all(entry, ec);
+
+        assert(!static_cast<bool>(ec));
+    }
+    //!\}
+
+    /*!\brief Removes a file or empty directory.
+     * \returns `true` if the  underlying path could be removed, otherwise `false`.
+     * \throws std::filesystem::filesystem_error on underlying OS API errors.
+     *
+     * \details
+     *
+     * Internally calls std::filesystem::remove on the stored std::filesystem::path.
+     */
+    bool remove()
+    {
+        return std::filesystem::remove(entry);
+    }
+
+    /*!\brief Removes a file or directory and all its contents, recursively.
+    * \returns `true` if the  underlying path could be removed, otherwise `false`.
+    * \throws std::filesystem::filesystem_error on underlying OS API errors.
+    *
+    * \details
+    *
+    * Internally calls std::filesystem::remove_all on the stored std::filesystem::path.
+    */
+    bool remove_all()
```

Comment:
Also remove_all() in the current fashion behaves differently from the destructor, because the destructor does not throw, but remove_all() does...

Reply:
Well the destructor should not throw. And that's why I added this member function to also have a throwable interface, because it can come handy in some places.